### PR TITLE
Implementation for Boavizta Cloud Impact Model 

### LINKED ISSUE #68: Create Boavizta SCI Cloud IMP
## Name 
bovizta.cloud.sci

## Scope
This model returns the SCI score and component E, I and M values for cloud providers as supported by the Boavizta API

## References
- https://doc.api.boavizta.org/Reference/cloud_route/

## Assumptions
N/A

##  Limitations
- The Boavizta API currently only supports a provider of AWS for this model.

## Validations

- List of allowed country codes: https://api.boavizta.org/docs#/utils/utils_get_all_countries_v1_utils_country_code_get
- List of supported providers: https://api.boavizta.org/docs#/cloud/server_get_all_provider_name_v1_cloud_all_providers_get
- List of supported instance_types: https://api.boavizta.org/docs#/cloud/server_get_all_archetype_name_v1_cloud_all_instances_get

## Inputs
- provider: The name of the cloud provider
- instance_type: The instance type of the VM offering on that cloud provider.
- usage_location: The location where the VM is based
- days_use_time: Fraction of a day the VM is used for.
- time_workload: Average CPU utilization for that duration **or** a breakdown of the CPU utilization for that duration in this format:
  - Array of {
        "time_percentage": % of time,
        "load_percentage": utilization for this time slice
      },
      {
        "time_percentage": % of time,
        "load_percentage": utilization for this time slice
      }
- verbose: true/false
- allocation: Method of calcualtion, linear or total

## Example IMPL

```yaml
- component: 
    model: 
      path: bovizta.cloud.sci
    config: 
      verbose: true
      allocation: total
      provider: aws
      instance_type: r6g.medium
    observations:
      shared:
        location: FRA  # usage_location
      series:
          -   datetime: 2023-07-06T00:00
              duration: 15 # translated into days_use_time
              cpu: 0.34  # translated into time_workload                          
          -   datetime: 2023-07-06T00:15
              duration: 15s                              
              cpu: 0.12                                    
          -   datetime: 2023-07-06T00:30
              duration: 15s                               
              cpu: 0.01                                     
          -   datetime: 2023-07-06T00:40
              duration: 15s                               
              cpu: 0.78            
```

## Observations

- Will be in the format of CPU utilization, durations.
- They additionally have a location which is either used to give more accurate embodied or if it's used for grid emissions we can ignore since we want to handle grid emissions ourselves.

### Multiple Observations

```yaml
series:
    -   datetime: 2023-07-06T00:00
        duration: 15s                             
        cpu: 0.34                                    
    -   datetime: 2023-07-06T00:15
        duration: 15s                              
        cpu: 0.12                                    
    -   datetime: 2023-07-06T00:30
        duration: 15s                               
        cpu: 0.01                                     
    -   datetime: 2023-07-06T00:40
        duration: 15s                               
        cpu: 0.78                                     
```

- We need to call the API multiple times, **once** per observation.

**Eventually**, we will need to convert the Energy numbers that are returned into Carbon with our own Grid Emissions data source that provides real-time grid emissions.

```yaml
series:
    -   datetime: 2023-07-06T00:00
        e: 123 # convert this to O with the I for this specific time and duration using own own data source of grid emissions
        m: 34               
    -   datetime: 2023-07-06T00:15
        e: 12 # convert this to O with the I for this specific time and duration
        m: 34                           
    -   datetime: 2023-07-06T00:30
        e: 13 # convert this to O with the I for this specific time and duration
        m: 34                            
    -   datetime: 2023-07-06T00:40
        e: 234 # convert this to O with the I for this specific time and duration
        m: 34                                 
```

## Outputs

This is the raw output from Boavizta

```json
{
  "gwp": {
    "manufacture": 970, # embodied (m) in gCO2eq
    "use": 2100,
    "unit": "kgCO2eq"
  },
  "pe": {
    "manufacture": 13000,
    "use": 71020, # energy (e) in kWh
    "unit": "MJ"
  },
  "adp": {
    "manufacture": 0.15,
    "use": 0.000354,
    "unit": "kgSbeq"
  }
}
```

ImpactMetric (SCI style) will look like so for **now**:

```
{
 'e': xxx
 'm': xxx
}
```

## Unit Tests

- Calling with observations calls the API and returns a value for **e** and **m**.
- Call with a single observation returns a single impact metric with the same timestamp and duration.
- Call with multiple observations returns a multiple impact metrics with the same timestamp and duration as the original observation.
- If you don't provide mandatory inputs it returns an error (please decide what the mandatory inputs are in terms of config and observations).
- When the Boavizta API doesn't work (400/500 whatever) it returns an error.
- If you provide an instance type that isn't supported, returns an error
- If you provide a provider that isn't supported returns an error.
- If you provide a region that isn't supported returns an error.

### PR DESCRIPTION
**Pull request by gnanakeethan**
- Accepts array of objects or a single object as input.
- Transforms the input from IMPL to Boavizta compatible data type
- Allows raw input for ease of use.
- Exposes "configure", "calculate" functions as expected by the IMPCON.

Closes #68

### CODE REVIEW

File: lib/boavizta.ts [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+    public name: string | undefined;
+    public verbose: boolean = false;
+    public allocation: string = "TOTAL";
+
+    modelIdentifier(): string {
+        return "org.boavizta.cloud.sci"
+    }
+
+    protected async captureStaticParams(staticParams: object) {
+        const staticParamsCast = staticParams;
```
Comment (jawache) [inline lib/boavizta.ts:202]: What's the reason to assign to a new variable and use the new variable for checks?

File: lib/boavizta.ts [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+        if ('verbose' in staticParamsCast) {
+            this.verbose = staticParamsCast.verbose as boolean ?? false;
+            staticParamsCast.verbose = undefined;
```
Comment (jawache) [inline lib/boavizta.ts:205]: Why assign to undefined?

File: lib/boavizta.ts [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+        }
+        if ('allocation' in staticParamsCast) {
+            const allocation = staticParamsCast.allocation as string ?? "TOTAL";
```
Comment (jawache) [inline lib/boavizta.ts:208]: I'm not aware of this syntax, what does ?? do here?

File: lib/boavizta.ts [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+        } else {
+            throw new Error("Improper configure: Missing instance_type parameter");
+        }
+    }
+
+    validateProvider(staticParamsCast: object) {
+        if (!('provider' in staticParamsCast)) {
+            throw new Error("Improper configure: Missing provider parameter");
+        } else {
+            const supportedProviders = ["aws"];
```
Comment (jawache) [inline lib/boavizta.ts:261]: There is a Boavizta endpoint to return supported providers

File: lib/boavizta.ts (hunk 1 above)
Comment (gnanakeethan) [reply lib/boavizta.ts:202]: It was used when I had casted the input to index signature for easier working with key value pairs. I will remove.

File: lib/boavizta.ts (hunk 2 above)
Comment (gnanakeethan) [reply lib/boavizta.ts:205]: These two configuration do not go in as a part of JSON body rather a URL query parameter. Therefore it is removed after capturing the value.

File: lib/boavizta.ts (hunk 3 above)
Comment (gnanakeethan) [reply lib/boavizta.ts:208]: null coalescing operator. it returns the second value if the first is undefined or null.

File: lib/boavizta.ts (hunk 4 above)
Comment (gnanakeethan) [reply lib/boavizta.ts]: Updated.

**Comment by jawache** [comment]
To fix the embodied issue let's add a configuration parameter called: expected-lifespan, with units of "years".
Then @gnanakeethan use the equation here https://github.com/Green-Software-Foundation/sci/blob/main/Software_Carbon_Intensity/Software_Carbon_Intensity_Specification.md#embodied-emissions to weight the embodied response from Boavizta's.

**Comment by jawache** [comment]
@gnanakeethan after some investigation I think we just need to use the LINEAR allocation method.

Linear means that this output changes with respect to the amount of time you input, so if you input more time the manufacturing carbon emissions reported increases.

```
    "gwp": {
      "manufacture": 0.086,
      "use": 0.01,
      "unit": "kgCO2eq"
    },
```

**Comment by jawache** [comment]
**Default**
For embodied carbon, use LINEAR and take manufacturing as is.
For energy default is to take it as is from the PE section.

**Configure It?**
- Perhaps explore doing our own embodied carbon attribution clacualtion from the raw data in here: https://api.boavizta.org/docs#/cloud/instance_cloud_impact_v1_cloud__post
- Perhaps returning secondary energy instead of primary energy by scaling the energy by some factor.

**Review by jawache — CHANGES_REQUESTED** [review]
@gnanakeethan remove the python folder for cleanliness, this isn't needed (we're implementing in another ticket and in TS)

**Comment by gnanakeethan** [comment]
@jawache I am unable to push to the branch

**Review by jawache — APPROVED** [review]
LGTM

**Comment by jawache** [comment]
@gnanakeethan can you please fix the conflicts and we can merge already with pre-approved from the project team.

**Comment by gnanakeethan** [comment]
@jawache Conflicts resolved via Rebasing.

**Comment by gnanakeethan** [comment]
@jawache I am merging this since it is already approved.

**Pull request merged**